File: sample-problems/AlgebraBlackBox/Genes/ConstantGene.ts
/*!
 * @author electricessence / https://github.com/electricessence/
 * Licensing: MIT https://github.com/electricessence/Genetic-Algorithm-Platform/blob/master/LICENSE.md
 */


import AlgebraGene from "../Gene";
import UnreducibleGene from "./UnreducibleGene";

const EMPTY = "";

export default class ConstantGene extends UnreducibleGene
{
	//protected
	toStringInternal():string
	{
		return this._multiple + EMPTY;
	}

	toStringContents():string
	{
		return EMPTY;
	}

	clone():ConstantGene
	{
		return new ConstantGene(this._multiple);
	}

	//noinspection JSUnusedLocalSymbols
	calculateWithoutMultiple(values:ArrayLike<number>):number
	{
		return 1;
	}

	equals(other:AlgebraGene):boolean
	{
		return other instanceof ConstantGene && this._multiple==other._multiple || super.equals(other);
	}
}

File: sample-problems/AlgebraBlackBox/Genes/ConstantGene.cs
using System.Threading.Tasks;

namespace AlgebraBlackBox.Genes
{
    public class ConstantGene : Gene
	{
		public ConstantGene(double multiple = 1) : base(multiple)
		{

		}


		public override string ToStringContents()
		{
			return string.Empty;
		}

		protected override string ToStringInternal()
		{
			return Multiple.ToString();
		}

		public new ConstantGene Clone()
		{
			return new ConstantGene(Multiple);
		}

        protected override GeneticAlgorithmPlatform.IGene CloneInternal()
        {
			return this.Clone();
        }

		static readonly Task<double> T1d = Task.FromResult(1d);
        protected override Task<double> CalculateWithoutMultipleAsync(double[] values)
        {
			return T1d;
        }

		protected override double CalculateWithoutMultiple(double[] values)
        {
			return 1d;
        }

		public bool Equivalent(IGene other)
		{
			return this==other || other is ConstantGene && Multiple == other.Multiple;
		}
    }
}
